src/data.ts

```diff
@@ -37,86 +37,122 @@ export interface Note {
 
-  const notebookIdsByName = new Map<string, string>();
-  notebooks.forEach((n) => notebookIdsByName.set(n.title, n.id));
-  const notebooksById = new Map<string, Notebook>();
-  notebooks.forEach((n) => notebooksById.set(n.id, n));
+type NotebookCache = {
+  notebookIdsByName: Map<string, NoteId>;
+  notebooksById: Map<NotebookId, Notebook>;
+};
 
-  // Get a list of valid notebook names to filter out.
-  filteredNotebookNames = filteredNotebookNames.filter((name) =>
-    notebookIdsByName.has(name)
-  );
+function emptyNotebookCache(): NotebookCache {
+  return {
+    notebookIdsByName: new Map(),
+    notebooksById: new Map(),
+  };
+}
 
-  function shouldIncludeNote(parent_id: string): boolean {
-    var parentNotebook: Notebook = notebooksById.get(parent_id);
-    // Filter out the direct parent.
-    if (filteredNotebookNames.includes(parentNotebook.title)) {
-      return isIncludeFilter;
-    }
+function shouldIncludeNotebook(
+  id: NotebookId,
+  {
+    filteredNotebookNames,
+    isIncludeFilter,
+    shouldFilterChildren,
+  }: NotebookFilter,
+  { notebooksById }: NotebookCache
+): boolean {
+  let parentNotebook: Notebook = notebooksById.get(id);
+  // Filter out the direct parent.
+  if (filteredNotebookNames.includes(parentNotebook.title)) {
+    return isIncludeFilter;
+  }
 
-    // Filter a note if any of its ancestor notebooks are filtered.
-    if (shouldFilterChildren) {
-      while (parentNotebook !== undefined) {
-        if (filteredNotebookNames.includes(parentNotebook.title)) {
-          return isIncludeFilter;
-        }
-        parentNotebook = notebooksById.get(parentNotebook.parent_id);
+  // Filter a note if any of its ancestor notebooks are filtered.
+  if (shouldFilterChildren) {
+    while (parentNotebook !== undefined) {
+      if (filteredNotebookNames.includes(parentNotebook.title)) {
+        return isIncludeFilter;
       }
+      parentNotebook = notebooksById.get(parentNotebook.parent_id);
     }
-    return !isIncludeFilter;
   }
+  return !isIncludeFilter;
+}
+
+function createNotebookCache(notebooks: Array<Notebook>): NotebookCache {
+  const cache = {
+    notebookIdsByName: new Map<string, NoteId>(),
+    notebooksById: new Map<NotebookId, Notebook>(),
+  };
+  notebooks.forEach((n) => cache.notebookIdsByName.set(n.title, n.id));
+  notebooks.forEach((n) => cache.notebooksById.set(n.id, n));
+  return cache;
+}
 
-  var filteredNotes = new Map<string, Note>();
+/**
+ * Returns a filtered map of notes by notebook name.
+ */
+export function filterNotesByNotebookName(
+  cache: NotebookCache,
+  filter: NotebookFilter,
+  notes: Map<NoteId, Note>
+) {
+  // No filtering needed.
+  if (filter.filteredNotebookNames.length < 1) return notes;
+
+  // Get a list of valid notebook names to filter out.
+  // TODO: Is this even needed?
+  //filter = {
+  //  ...filter,
+  //  filteredNotebookNames: filter.filteredNotebookNames.filter((name) =>
+  //    cache.notebookIdsByName.has(name)
+  //  ),
+  //};
```

Comment:
Please evaluate if needed

Reply:
I vaguely remember running into some sort of exception with invalid notebook names.  I wasn't able to reproduce it locally with your PR.  Maybe it's safe to delete?